test/remote/recovery_code_tests.js

```diff
@@ -0,0 +1,120 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+'use strict'
+
+const assert = require('insist')
+const config = require('../../config').getProperties()
+const TestServer = require('../test_server')
+const Client = require('../client')()
+const otplib = require('otplib')
+
+describe('remote totp', function () {
+  let server, client, email, recoveryCodes
+  const password = 'pssssst'
+  const metricsContext = {
+    flowBeginTime: Date.now(),
+    flowId: '0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef'
+  }
+
+  this.timeout(10000)
+
+  otplib.authenticator.options = {
+    encoding: 'hex',
+    step: config.step
+  }
+
+  before(() => {
+    return TestServer.start(config)
+      .then(s => {
+        server = s
+      })
+  })
+
+  beforeEach(() => {
+    email = server.uniqueEmail()
+    return Client.createAndVerify(config.publicUrl, email, password, server.mailbox)
+      .then((x) => {
+        client = x
+        assert.ok(client.authAt, 'authAt was set')
+        return client.createTotpToken({metricsContext})
+          .then((result) => {
+            otplib.authenticator.options = {
+              secret: result.secret
+            }
+
+            // Verify TOTP token so that initial recovery codes are generated
+            const code = otplib.authenticator.generate()
+            return client.verifyTotpCode(code, {metricsContext})
+              .then((response) => {
+                assert.equal(response.success, true, 'totp codes match')
+
+                recoveryCodes = response.recoveryCodes
+                assert.equal(response.recoveryCodes.length, 8, 'recovery codes returned')
+                return server.mailbox.waitForEmail(email)
+              })
+              .then((emailData) => {
+                assert.equal(emailData.headers['x-template-name'], 'postAddTwoStepAuthenticationEmail', 'correct template sent')
+              })
+          })
+      })
+  })
+
+  it('should create recovery codes', () => {
+    assert.ok(recoveryCodes)
+    assert.equal(recoveryCodes.length, 8, 'recovery codes returned')
```

Comment:
Should we also assert the content of `recoveryCodes`, e.g. that the strings are the expected length, only contain hex, whatever?